honeybadgermpc/hbavss.py

```diff
@@ -229,57 +232,144 @@ def __exit__(self, typ, value, traceback):
+        r1_set = set()
+        r2_set = set()
+        ready_sent = False
+        r1_sent = False
+        r2_sent = False
+        r1_phi = [None] * self.n
+        r2_phi = [None] * self.n
 
-        # if 2t+1 ready -> output shares
-        while len(ready_set) < 2 * self.t + 1:
-            sender, avss_msg = await recv()  # First value is the `sid`
-            if avss_msg == HbAVSSMessageType.READY and sender not in ready_set:
+        while True:
+            # main recv loop for Bracha-style agreement and implicate handling
+            sender, avss_msg = await recv()
+            # OK
+            if avss_msg[0] == HbAVSSMessageType.OK and sender not in ok_set:
+                ok_set.add(sender)
+            # READY
+            if avss_msg[0] == HbAVSSMessageType.READY and sender not in ready_set:
                 ready_set.add(sender)
-
-        return shares
+            # IMPLICATE
+            if avss_msg[0] == HbAVSSMessageType.IMPLICATE:
+                # Handling Implication
+                j = sender
+                j_pk = avss_msg[1]
+                j_k = avss_msg[2]
+                # discard if PKj ! = g^SKj
+                if self.public_keys[j] != pow(self.g, j_pk):
+                    logger.debug("[%d] invalid implicate from sender %s",
+                                 self.my_id, sender)
+                    continue
+                # decrypt and verify
+                implicate_msg = await avid.retrieve(tag, j)
+                j_z = loads(implicate_msg)
+                j_shared_key = pow(ephemeral_public_key, j_pk)
+                j_share, j_aux, j_witnesses = SymmetricCrypto.decrypt(
+                    str(j_shared_key).encode("utf-8"), j_z[j_k])
+                if self.poly_commit.verify_eval(
+                        commitments[j_k], sender+1, j_share, j_aux, j_witnesses):
+                    continue
+
+                # proceed to share recovery
+                logger.debug("[%d] Share recovery activated by %d", self.my_id, j)
+                # setup the coordinates for interpolation
+                c_coords = [None] * secret_size
+                phi_coords = [None] * secret_size
+                aux_coords = [None] * secret_size
+                w_coords = [None] * secret_size
+                for i in range(secret_size):
+                    c_coords[i] = (i, commitments[i])
+                    phi_coords[i] = (i, shares[i])
+                    aux_coords[i] = (i, auxes[i])
+                    w_coords[i] = (i, witnesses[i])
+                # interpolate commitments
+                interpolated_c = [None] * self.n
+                for i in range(self.n):
+                    interpolated_c[i] = interpolate_g1_at_x(c_coords, i)
+                # if my shares are valid
+                if not r1_sent and self.shares_future.done():
+                    r1_sent = True
+                    for j in range(self.n):
+                        phi_i_j = interpolate_at_x(phi_coords, j)
```

Comment:
same here, this looks like you are interpolating at all `n` points - consider using Vandermonde/FFT

Reply:
As this is ZR, FFT should be used, yeah